lib/srv/authhandlers_test.go

```diff
@@ -225,3 +233,177 @@ func TestRBAC(t *testing.T) {
+		ActiveKeys: types.CAKeySet{
+			SSH: []*types.SSHKeyPair{
+				{
+					PublicKey:      userCAPriv.MarshalSSHPublicKey(),
+					PrivateKey:     userCAPriv.PrivateKeyPEM(),
+					PrivateKeyType: types.PrivateKeyType_RAW,
+				},
+			},
+		},
+	})
+	require.NoError(t, err)
+
+	// create mock SSH server and add a cluster name
+	server := newMockServer(t)
+	cn, err := types.NewClusterName(types.ClusterNameSpecV2{
+		ClusterName: clusterName,
+		ClusterID:   "cluster_id",
+	})
+	require.NoError(t, err)
+	err = server.auth.SetClusterName(cn)
+	require.NoError(t, err)
+	ctx := context.Background()
+
+	accessPoint := &mockCAandAuthPrefGetter{
+		AccessPoint: server.auth,
+		cas: map[types.CertAuthType]types.CertAuthority{
+			types.UserCA: userCA,
+		},
+	}
+
+	// create auth handler and dummy node
+	config := &AuthHandlerConfig{
+		Server:      server,
+		Emitter:     &eventstest.MockRecorderEmitter{},
+		AccessPoint: accessPoint,
+	}
+	ah, err := NewAuthHandlers(config)
+	require.NoError(t, err)
+
+	node, err := types.NewServer("testie_node", types.KindNode, types.ServerSpecV2{
+		Addr:     "1.2.3.4:22",
+		Hostname: "testie",
+		Version:  types.V2,
+	})
+	require.NoError(t, err)
+
+	mfaAuthPref, err := types.NewAuthPreference(types.AuthPreferenceSpecV2{
+		SecondFactor:   constants.SecondFactorOTP,
+		RequireMFAType: types.RequireMFAType_HARDWARE_KEY_TOUCH,
+	})
+	require.NoError(t, err)
+
+	noMFAAuthPref, err := types.NewAuthPreference(types.AuthPreferenceSpecV2{
+		SecondFactor:   constants.SecondFactorOTP,
+		RequireMFAType: types.RequireMFAType_OFF,
+	})
+	require.NoError(t, err)
+
+	// create roles
+	joinMFARole, err := types.NewRole("joinMFA", types.RoleSpecV6{
+		Options: types.RoleOptions{
+			RequireMFAType: types.RequireMFAType_SESSION,
+		},
+		Allow: types.RoleConditions{
+			NodeLabels: types.Labels{
+				types.Wildcard: []string{types.Wildcard},
+			},
+		},
+	})
+	require.NoError(t, err)
+	_, err = server.auth.CreateRole(ctx, joinMFARole)
+	require.NoError(t, err)
+
+	joinRole, err := types.NewRole("join", types.RoleSpecV6{
+		Allow: types.RoleConditions{
+			NodeLabels: types.Labels{
+				types.Wildcard: []string{types.Wildcard},
+			},
+		},
+	})
+	require.NoError(t, err)
+	_, err = server.auth.CreateRole(ctx, joinRole)
+	require.NoError(t, err)
+
+	tests := []struct {
+		name      string
+		authPref  types.AuthPreference
+		role      string
+		testError func(t *testing.T, err error)
+	}{
+		{
+			name:     "MFA cluster auth, MFA role",
+			authPref: mfaAuthPref,
+			role:     joinMFARole.GetName(),
+			testError: func(t *testing.T, err error) {
+				require.Error(t, err)
+				require.True(t, trace.IsAccessDenied(err))
+			},
+		},
+		{
+			name:     "MFA cluster auth, no MFA role",
+			authPref: mfaAuthPref,
+			role:     joinRole.GetName(),
+			testError: func(t *testing.T, err error) {
+				require.Error(t, err)
+				require.True(t, trace.IsAccessDenied(err))
+			},
+		},
+		{
+			name:     "no MFA cluster auth, MFA role",
```

Comment:
I kind of expected this one to fail, is the role not a factor in this check? Could we maybe cut the roles from the test?

If the role is indeed a factor here, could you add a brief comment explaining why this test case is not expected to error?

Reply:
I expected this to fail as well, but I thought the cluster auth MFA setting must override the session MFA setting. Examining further proved this not to be the case, and this test case passed because of the `join` role always getting added to the SSH certificate in the test rather than what was specified in the test case. The auth logic was sound, only the test was updated to fix this.